### ISSUE TITLE
Add CustomListener to Opts

### ISSUE THREAD

**Issue by hansonkd** [nats-io/gnatsd#668]
- [ ] Defect
 - [x] Feature Request or Change Proposal
 
## Feature Requests
Allow passing an initialized `net.Listner` so when the Server starts, it does not bind a new connection. This gives more control over the socket.

#### Use Case:
This would allow more advanced custom handshakes for TCP connections. Paired with CustomDialer in `go-nats` it would allow you to easily setup custom connection initialization. This is useful for more advanced security protocols or if you want to run multiple servers on one connection.
 
#### Proposed Change:
* Add `CustomListener` option to Opts. 
* In `AcceptLoop` use `Opts.CustomListner` if available else fallback to `net.Listen`.
 
#### Who Benefits From The Change(s)?
People  who want more control over how connections are handled.

**Comment by hansonkd** [comment]
I would also like to point out that passing in a Listener would allow us to take full advantage of the golang `net.Conn` interface. For websockets, we would no longer have to have a separate proxy, but rather run the WS server along side of the nats server and pass the `Conn` directly (no extra TCP connections, buffering and piping).

**Comment by derekcollison** [comment]
Can you describe more some of the custom security setup you are trying to achieve? Also for WS the architecture will be very different, with different memory and performance profiles than a normal TCP/IP connection.

**Comment by derekcollison** [comment]
@hansonkd Could I get a bit more information on this one? Want to close or develop a plan. Thanks.

**Comment by ripienaar** [comment]
I have not tried but I was keeping an eye on this one too as I need to achieve FIPS compliance where you have to link against a specific SSL library - which NATS doesn't support.

I could take care of setting up the connection using this compliant library and pass it in.  

But as I said I have only been keeping an eye on this not actually explored if this would solve my problem but it seems useful

**Comment by ripienaar** [comment]
@derekcollison exactly, except I'd set up the connection ahead of time and pass it in.  In effect it acts like a tunnel but one users don't need to manage.

**Comment by derekcollison** [comment]
Pass it in to a client? Currently the golang TLS does not (IIRC) support FIPS, so this would need to be external, but could be setup on a well known local port, etc.

**Comment by ripienaar** [comment]
It's a load of bullshit but there's a thing called BoringSSL that lets you do this for Go via a bunch of CLANG flags at compile time, gov will validate that a specific hash of this lib can be used in FIPS compliant software.

One would either set up a managed tunnel or arrange for ones tools to take a prepared socket.

Again, I've not tried this only been doing some light research as I can see this train coming :(

**Comment by ripienaar** [comment]
And to be clear, it would need to be used then for ALL things including cluster comms etc, I dont really realistically see being able to pass sockets in as a solution here - instead I think I'll have to build these managed tunnels using this library and make NATS speak clear over it.

Anyway, just thought I'd throw in a passing comment on this for some where this MIGHT be useful

**Comment by derekcollison** [comment]
Very helpful, thanks..
